lib/ansible/modules/cloud/cloudstack/cs_instance_password_reset.py

```diff
@@ -0,0 +1,163 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible. If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: cs_instance_password_reset
+short_description: Allows resetting VM the default passwords on Apache CloudStack based clouds.
+description:
+    - Resets the default user account's password on an instance.
+    - Requires cloud-init to be installed in the virtual machine.
+    - The passwordenabled flag must be set on the template associated with the VM.
+version_added: '2.8'
+author: "Gregor Riepl (@onitake)"
+options:
+  vm:
+    description:
+      - Name of the virtual machine to reset the password on.
+    required: true
+  domain:
+    description:
+      - Name of the domain the virtual machine belongs to.
+  account:
+    description:
+      - Account the virtual machine belongs to.
+  project:
+    description:
+      - Name of the project the virtual machine belongs to.
+  zone:
+    description:
+      - Name of the zone in which the instance is deployed.
+      - If not set, the default zone is used.
+  poll_async:
+    description:
+      - Poll async jobs until job has finished.
+    default: yes
+    type: bool
+extends_documentation_fragment: cloudstack
+'''
+
+EXAMPLES = '''
+- name: reset and get the root password
+  local_action:
+    module: cs_instance_password_reset
+    name: myvirtualmachine
+  register: root
+- debug:
+    msg: "new root password is {{ root.password }}"
+- name: reboot the virtual machine to activate the new password
+  local_action:
+    module: cs_instance
+    name: myvirtualmachine
+    state: restarted
+  when: root is changed
+'''
+
+RETURN = '''
+---
+id:
+  description: ID of the virtual machine.
+  returned: success
+  type: string
+  sample: a6f7a5fc-43f8-11e5-a151-feff819cdc9f
+password:
+  description: The new default password.
+  returned: success
+  type: string
+  sample: ahQu5nuNge3keesh
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.cloudstack import (
+    AnsibleCloudStack,
+    cs_required_together,
+    cs_argument_spec
+)
+
+
+class AnsibleCloudStackPasswordReset(AnsibleCloudStack):
+
+    def __init__(self, module):
+        super(AnsibleCloudStackPasswordReset, self).__init__(module)
+        self.returns = {
+            'password': 'password',
+        }
+        self.password = None
+
+    def reset_password(self):
+        args = {
+            'domainid': self.get_domain(key='id'),
+            'account': self.get_account(key='name'),
+            'projectid': self.get_project(key='id'),
+            'zoneid': self.get_zone(key='id'),
+            'id': self.get_vm(key='id'),
```

Comment:
I think `id` is the only required argument here since the cs api call only takes `id` as argument. The common Cloudstack library will return the correct id of the vm based on other args.

Reply:
Cross-checked with cs_instance_nic, and changed.